File: nicegui/elements/plotly/plotly.js
import { convertDynamicProperties } from "../../static/utils/dynamic_properties.js";

const bundles = {};
const traceTypes = new WeakMap();

function tracesOf(value) {
  if (Array.isArray(value)) return value;
  if (value?.data || value?.frames) {
    return [...(value.data ?? []), ...(value.frames ?? []).flatMap((frame) => frame?.data ?? [])];
  }
  return [value];
}

function loadBundle(full) {
  const specifier = full ? "nicegui-plotly/full.js" : "nicegui-plotly";
  bundles[specifier] ??= import(specifier).then(({ Plotly }) => {
    traceTypes.set(Plotly, new Set(Object.keys(Plotly.PlotSchema.get().traces)));
    return Plotly;
  });
  return bundles[specifier];
}

export default {
  template: "<div></div>",
  async mounted() {
    this.full_loaded = this.full;
    this.Plotly = await loadBundle(this.full);
    this.update();
    this.$nextTick(() => {
      this.resizeObserver = new ResizeObserver(() => {
        if (this.options.config?.responsive === false) return;
        this.Plotly.Plots.resize(this.$el);
      });
      this.resizeObserver.observe(this.$el);
    });
  },
  unmounted() {
    this.resizeObserver?.disconnect();
  },
  methods: {
    update() {
      // wait for plotly to be loaded
      if (typeof this.Plotly === "undefined") {
        setTimeout(this.update, 10);
        return;
      }

      // swap in the full bundle if the figure uses a trace type the loaded one does not know
      if (this.needs_full_bundle(this.options)) {
        this.upgrade().then(this.update);
        return;
      }

      // default responsive to true
      const options = this.options;
      if (options.config?.responsive === true) options.config.responsive = undefined;

      // reuse plotly instance if config is the same
      let promise;
      if (this.last_options && JSON.stringify(options.config) === JSON.stringify(this.last_options.config)) {
        promise = this.Plotly.react(this.$el, this.options, options.config);
      } else {
        promise = this.Plotly.newPlot(this.$el, this.options, options.config);
        this.set_handlers();
      }

      // store last options
      this.last_options = options;
      return promise;
    },
    needs_full_bundle(figure) {
      // an unregistered trace type is not an error in plotly.js, it silently renders as a scatter trace
      if (this.full_loaded) return false; // NOTE: a type the full bundle does not know either must not loop
      const known = traceTypes.get(this.Plotly);
      return tracesOf(figure).some((trace) => trace?.type !== undefined && !known.has(trace.type));
    },
    upgrade() {
      this.upgrading ??= (async () => {
        const light = this.Plotly;
        this.Plotly = await loadBundle(true);
        this.full_loaded = true;
        if (this.last_options) light.purge(this.$el); // the light bundle's plot state is not the full one's to reuse
        this.last_options = null;
      })();
      return this.upgrading;
    },
    async run_plot_method(name, ...args) {
      if (typeof this.Plotly === "undefined") {
        logAndEmit("error", "Plotly is not loaded yet.");
        return;
      }
      // methods taking trace data take it as their first argument, so that is the only one worth checking
      if (this.needs_full_bundle(args[0])) {
        await this.upgrade();
        await this.update();
      }
      convertDynamicProperties(args, true);
      const result = await runMethod(this.Plotly, name, [this.$el, ...args]);
      // most plotly.js functions resolve to the graph element, which cannot be serialized back to the server
      return result === this.$el ? undefined : result;
    },
    set_handlers() {
      // forward events
      for (const name of [
        // source: https://plotly.com/javascript/plotlyjs-events/
        "plotly_click",
        "plotly_legendclick",
        "plotly_selecting",
        "plotly_selected",
        "plotly_hover",
        "plotly_unhover",
        "plotly_legenddoubleclick",
        "plotly_restyle",
        "plotly_relayout",
        "plotly_webglcontextlost",
        "plotly_afterplot",
        "plotly_autosize",
        "plotly_deselect",
        "plotly_doubleclick",
        "plotly_redraw",
        "plotly_animated",
      ]) {
        this.$el.on(name, (event) => {
          const args = {
            ...event,
            points: event?.points?.map((p) => ({
              ...p,
              fullData: undefined,
              xaxis: undefined,
              yaxis: undefined,
            })),
            xaxes: undefined,
            yaxes: undefined,
          };
          this.$emit(name, args);
        });
      }
    },
  },
  data() {
    return {
      last_options: null,
      upgrading: null,
      full_loaded: false,
    };
  },
  props: {
    options: Object,
    full: Boolean,
  },
};

File: nicegui/elements/plotly/plotly.py
from __future__ import annotations

from contextlib import suppress

from ... import optional_features
from ...awaitable_response import AwaitableResponse
from ...element import Element

with suppress(ImportError):
    import plotly.graph_objects as go
    optional_features.register('plotly')

LIGHT_TRACE_TYPES = {  # trace types registered in dist/index.js, see src/index.mjs
    'bar', 'box', 'contour', 'heatmap', 'histogram', 'histogram2d', 'histogram2dcontour',
    'image', 'indicator', 'pie', 'scatter', 'table', 'violin',
}


class Plotly(Element, component='plotly.js', esm={'nicegui-plotly': 'dist'}):

    def __init__(self, figure: dict | go.Figure) -> None:
        """Plotly Element

        Renders a Plotly chart.
        There are two ways to pass a Plotly figure for rendering, see parameter `figure`:

        * Pass a `go.Figure` object, see https://plotly.com/python/

        * Pass a Python `dict` object with keys `data`, `layout`, `config` (optional), see https://plotly.com/javascript/

        For best performance, use the declarative `dict` approach for creating a Plotly chart.

        Figures using only common trace types load a light plotly.js bundle roughly a third the size of the full one.
        The full bundle is fetched instead as soon as a figure needs a trace type the light bundle does not contain.

        :param figure: Plotly figure to be rendered. Can be either a `go.Figure` instance, or
                       a `dict` object with keys `data`, `layout`, `config` (optional).
        """
        super().__init__()

        self.figure = figure
        self.update()
        self._classes.append('js-plotly-plot')
        self._update_method = 'update'

    def update_figure(self, figure: dict | go.Figure):
        """Overrides figure instance of this Plotly chart and updates chart on client side."""
        self.figure = figure
        self.update()

    def run_plot_method(self, name: str, *args, timeout: float = 1) -> AwaitableResponse:
        """Run a plotly.js library function against the chart's HTML element.

        See the `plotly.js function reference <https://plotly.com/javascript/plotlyjs-function-reference/>`_ for a list of methods.
        The chart's HTML element is passed as the first argument automatically.

        If the function is awaited, the result of the method call is returned
        (unless it resolves to the chart's HTML element, in which case ``None`` is returned).
        Otherwise, the method is executed without waiting for a response.

        *Added in version 3.13.0*

        :param name: name of the plotly.js function (without the ``Plotly.`` prefix)
        :param args: arguments to pass after the chart element
        :param timeout: timeout in seconds (default: 1 second)

        :return: AwaitableResponse that can be awaited to get the result of the method call
        """
        return self.run_method('run_plot_method', name, *args, timeout=timeout)

    def update(self) -> None:
        with self._props.suspend_updates():
            self._props['options'] = self._get_figure_json()
            if not self._props.get('full'):  # NOTE: never switch back, the client has loaded the full bundle already
                self._props['full'] = self._needs_full_bundle()
        super().update()

    def _needs_full_bundle(self) -> bool:
        options = self._props['options']
        try:
            traces = [*(options.get('data') or []),
                      *(trace for frame in options.get('frames') or [] for trace in frame.get('data') or [])]
        except (AttributeError, TypeError):
            return True  # an unexpected figure shape is not worth guessing about
        for trace in traces:
            try:
                trace_type = trace['type'] or 'scatter'
            except (KeyError, TypeError):
                trace_type = 'scatter'
            if trace_type not in LIGHT_TRACE_TYPES:
                return True
        return False

    def _get_figure_json(self) -> dict:
        if optional_features.has('plotly') and isinstance(self.figure, go.Figure):
            # convert go.Figure to dict object which is directly JSON serializable
            # orjson supports NumPy array serialization
            return self.figure.to_plotly_json()

        if isinstance(self.figure, dict):
            # already a dict object with keys: data, layout, config (optional)
            return self.figure

        raise ValueError(f'Plotly figure is of unknown type "{self.figure.__class__.__name__}".')
